ecs-agent/api/ecs/client/ecs_client_factory.go

```diff
@@ -0,0 +1,61 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package ecsclient
+
+import (
+	"github.com/aws/amazon-ecs-agent/ecs-agent/api/ecs"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/config"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/ec2"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+)
+
+// ECSClientFactory interface can be used to create new ECS clients.
+// It wraps the internal ecsClientFactory and can help ease writing unit test code for consumers.
+type ECSClientFactory interface {
+	// NewClient creates a new ECS client.
+	NewClient() (ecs.ECSClient, error)
+	// Credentials returns the credentials chain used by the ECS client.
+	Credentials() *credentials.Credentials
+}
+
+type ecsClientFactory struct {
+	credentialsProvider *credentials.Credentials
+	configAccessor      config.AgentConfigAccessor
+	ec2MetadataClient   ec2.EC2MetadataClient
+	agentVer            string
```

Comment:
nit: I don't see the point of shortening this variable name, could we use `agentVersion` instead?

Also, could this be gleaned from a ConfigAccessor instead? Why does this need to be in the factory struct?

Reply:
> nit: I don't see the point of shortening this variable name, could we use agentVersion instead?

Sure, if you prefer I can change the variable name `agentVer` to `agentVersion` as part of any future revision addressing blocking changes.

> Also, could this be gleaned from a ConfigAccessor instead? Why does this need to be in the factory struct?

This needs to be in the factory struct because it is currently a function parameter of the `NewECSClient` function. It may be possible to refactor the ConfigAccessor interface and its implementers to take on a new method returning Agent version, but that's out of scope of this pull request.